src/ncp/spinel_transformer.hpp

```diff
@@ -0,0 +1,35 @@
+/**
+ * NCP Spinel Transformer allows to transform spinel frames sent between Application Processor (AP) and Network Co-Processor (NCP).
+ */
+
+namespace SpinelTransformer {
+
+/**
+ * Transforms spinel frames before sending to AP/NCP.
+ *
+ * This method transforms outbound frames in both directions, i.e. from AP to NCP and from NCP to AP.
+ * NOTE: \p aOutputLength might be different than \p aInputLength, so make sure to allocate enough memory.
+ *
+ * @param[in] aInput Pointer to buffer containing spinel frame.
+ * @param[in] aInputLength Length of spinel frame.
+ * @param[out] aOutput Pointer to allocated buffer for transformed spinel frame.
+ * @param[in,out] aOutputLength Length of allocated buffer for transformed spinel frame.
+ * In return, it will contain the actual length of transformed frame.
+ * @return \c true on success, \c false otherwise.
+ */
+bool TransformOutbound(const unsigned char *aInput, size_t aInputLength, unsigned char *aOutput, size_t *aOutputLength);
```

Comment:
Wouldn't a proper security layer need more methods than just `TransformOutbound/TransformInbound`? For example, how would both sides agree on their respective frame counter values?

Perhaps I'm thinking too much about security in cases where you would want to maintain usability in the event that someone is actively trying to attack a secure connection. I'm guessing that the devices getting out of sync and no longer communicating when under active attack would be acceptable for a channel that is intended to be a physically isolated direct connection. In other words, (and thinking out loud) I guess you don't need to protect against attacks which might cause DoS, because if this sort of attack is happen there are easier ways to cause DoS. And if you don't need to protect against those sorts of attacks, then there shouldn't be too much harm in making the protocol self-synchronize.

So I guess I answered my own question. Heh. I'll leave this comment here in case you'd like to elaborate on what kind of security mechanism you were imagining.

Reply:
Right. We mostly were concerned about clear-text aspect. As for going out of sync - I guess the code with this modification is not worse than the original. I.e. what will happen if plain spinel will lose some frame? Probably exactly the same thing will happen with the change. Active attack against the connection indeed roughly equals DoS situation. As long as attacker does not get elevated privileges (as it could be the case with plaintext) and can not read the traffic, I'm ok with it. If attacker has physical access to the device, he/she can smash it with the hammer as well, so it is really hard to protect from this kind of DoS in software ;)

That being said, we'll see if there is anything sensible for us to help with self-synchronization,  maybe there are some low hanging fruits there.